From a1e7912016dbd9a877f8a6f0107d26830269bb34 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 6 Nov 2018 15:32:20 +0100
Subject: [PATCH 9/9] [SQUSH][DNM] Merge tag 'android-9.0.0_r16' into
 lineage-16.0

commit ecc27018512ab452192cf074304d6b10ea41b70e
Merge: 81abfa2c 17059fe5
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Aug 8 03:15:27 2018 +0000

    Snap for 4939496 from 17059fe53b7f535886918d06fea58a90930862fc to pi-qpr1-release

    Change-Id: Id8422a6cb43aa012ff53dd70a684b3b3c859ec51

commit 17059fe53b7f535886918d06fea58a90930862fc
Author: Greg Kaiser <gkaiser@google.com>
Date:   Wed Aug 1 13:15:19 2018 -0700

    cryptfs: Remove Speck support

    Remove the Speck encryption support. It was eventually
    decided not to allow Speck in Android P, so this code
    is no longer needed and wasn't used outside of testing.

    Note we don't just "git revert" the original commit
    (38723f23ff521e95ed295f500e6529c52a994c9f) because we want
    to retain the infrastructure for allowing new types of
    crypto algorithms in the future.

    Bug: 112009351
    Test: Attempted to setup a device with ro.crypto.fde_algorithm set to Speck, and Speck was rejected and the system defaulted to AES.
    Change-Id: I69a8b4e8632f8d30b5b54783cb986ab42d4397d9

commit 81abfa2ca6044c01f06ae5dd32a2cb66938aebae
Merge: 8e7c2624 31e962fe
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Jul 31 03:11:07 2018 +0000

    Snap for 4922203 from 31e962fe1970b12ec216cf3e376d126d082144a4 to pi-qpr1-release

    Change-Id: Ia64205e7135bcd0ced71c3e82fe59ffeb23d760a

commit 31e962fe1970b12ec216cf3e376d126d082144a4
Author: Jaegeuk Kim <jaegeuk@google.com>
Date:   Sun Jul 29 06:56:57 2018 -0700

    vold: meta encryption: fix /dev/block/by-name/userdata to dm-3

    This fixes F2FS GC failure in idle-maint.
    07-28 18:25:54.838   603 11187 D vold    : idle maintenance started
    07-28 18:25:54.846   603 11187 D vold    : Start GC on /sys/fs/f2fs/sda21
    07-28 18:25:54.847   603 11187 W vold    : Set discard gralunarity failed on/sys/fs/f2fs/sda21: No such file or directory
    07-28 18:25:54.847   603 11187 W vold    : Start GC failed on /sys/fs/f2fs/sda21: No such file or directory

    Bug: 111953875
    Change-Id: I1d10802121d5641cf9ba780fee249affd2cf6ffe
    Signed-off-by: Jaegeuk Kim <jaegeuk@google.com>

Change-Id: I520caf85a5bd025b75a3984627f1a630531e9edd
---
 IdleMaint.cpp         | 8 +++++---
 model/PrivateVolume.h | 1 +
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/IdleMaint.cpp b/IdleMaint.cpp
index 7744024..5a19b8c 100644
--- a/IdleMaint.cpp
+++ b/IdleMaint.cpp
@@ -15,6 +15,7 @@
  */
 
 #include "IdleMaint.h"
+#include "FileDeviceUtils.h"
 #include "Utils.h"
 #include "VolumeManager.h"
 #include "model/PrivateVolume.h"
@@ -85,8 +86,8 @@ static void addFromVolumeManager(std::list<std::string>* paths,
             } else if (path_type == PathTypes::kBlkDevice) {
                 std::string gc_path;
                 const std::string& fs_type = vol->getFsType();
-                if (fs_type == "f2fs" &&
-                    Realpath(vol->getRawDevPath(), &gc_path)) {
+                if (fs_type == "f2fs" && (Realpath(vol->getRawDmDevPath(), &gc_path) ||
+                                          Realpath(vol->getRawDevPath(), &gc_path))) {
                     paths->push_back(std::string("/sys/fs/") + fs_type +
                                      "/" + Basename(gc_path));
                 }
@@ -130,7 +131,8 @@ static void addFromFstab(std::list<std::string>* paths, PathTypes path_type) {
         } else if (path_type == PathTypes::kBlkDevice) {
             std::string gc_path;
             if (std::string(fstab->recs[i].fs_type) == "f2fs" &&
-                Realpath(fstab->recs[i].blk_device, &gc_path)) {
+                Realpath(android::vold::BlockDeviceForPath(
+                    std::string(fstab->recs[i].mount_point) + "/"), &gc_path)) {
                 paths->push_back(std::string("/sys/fs/") + fstab->recs[i].fs_type +
                                  "/" + Basename(gc_path));
             }
diff --git a/model/PrivateVolume.h b/model/PrivateVolume.h
index 9508671..9a61f8d 100644
--- a/model/PrivateVolume.h
+++ b/model/PrivateVolume.h
@@ -41,6 +41,7 @@ public:
     virtual ~PrivateVolume();
     const std::string& getFsType() { return mFsType; };
     const std::string& getRawDevPath() { return mRawDevPath; };
+    const std::string& getRawDmDevPath() { return mDmDevPath; };
 
 protected:
     status_t doCreate() override;
-- 
2.17.1

